### PR TITLE
Issue #594 Adding explicit clean_docs target in Makefile. Refactored …

### LINKED ISSUE #594: Permission denied error on `make clean` for `docs/build` dir content
Hi,

On performing `make clean`, I am getting permission denied error for content of `docs/build` dir:
```
$ make clean
rm -rf /home/budhram/gowork/pkg/linux_amd64/github.com/minishift
rm -rf ./out
rm -rf release
rm -rf vendor
rm -rf ./docs/build
rm: cannot remove './docs/build/minishift-adoc.tar': Permission denied
rm: cannot remove './docs/build/_topic_map.yml': Permission denied
rm: cannot remove './docs/build/variables.adoc': Permission denied
Makefile:107: recipe for target 'clean' failed
make: *** [clean] Error 1


#### ls info of docs/build
$ ls docs/build -l
total 120
-rw-r--r--. 1 root root 112640 Mar 20 20:04 minishift-adoc.tar
-rw-r--r--. 1 root root   2297 Mar 20 20:04 _topic_map.yml
-rw-r--r--. 1 root root     32 Mar 20 20:04 variables.adoc
```
It seems those content are created with `root` user permission.

Interestingly, I thought `rake clean` inside `docs` folder should clean it up but it seems to be not working too
```
$ pwd
/home/budhram/gowork/src/github.com/minishift/minishift/docs

$ ls
build  config.rb  Dockerfile  Gemfile  Gemfile.lock  Rakefile  source

$ tree build 
build
├── minishift-adoc.tar
├── _topic_map.yml
└── variables.adoc

0 directories, 3 files

$ rake clean
rm -f 
rm -rf build

$ tree build 
build
├── minishift-adoc.tar
├── _topic_map.yml
└── variables.adoc

0 directories, 3 files
```

At the end, I need to `sudo rm -rf docs/build`

### PR DESCRIPTION
**Pull request by hferentschik**
Fix #594 

…Dockerfile to create docs user

### CODE REVIEW

**Comment by hferentschik** [comment]
@budhrg have a look whether this fixes your problem. Not, _make clean_ will definitely work now, since I am not deleting the docs/build directory there anymore. It was not the right place to begin with. Instead I added a dedicated _clean_docs_ target. 

I also went ahead and refactored the Dockerfile a bit, avoiding the root user creation.

**Comment by hferentschik** [comment]
@LalatenduMohanty did you try it?
@budhrg Any chance you verified this pull request yet?

**Comment by budhrg** [comment]
> @budhrg Any chance you verified this pull request yet?

No @hferentschik , didn't get chance. Will do after some time.

**Comment by budhrg** [comment]
@hferentschik  Will do the review after some time. If looks ok then merge it.

**Comment by hferentschik** [comment]
@budhrg 

> If looks ok then merge it.

please do

**Comment by budhrg** [comment]
Hi @hferentschik ,

Unfortunately, I am not able to pass `make gen_docs` command. `make clean_docs` seems to be working now.

```
...
Step 15/15 : CMD -T
 ---> Using cache
 ---> eae1e8d53293
Successfully built eae1e8d53293
cd docs && docker run -e MINISHIFT_VERSION=1.0.0-beta.5 -tiv /home/budhram/gowork/src/github.com/minishift/minishift/docs:/home/docs:Z minishift/docs gen
mkdir -p build
rake aborted!
Errno::EACCES: Permission denied @ dir_s_mkdir - build
/home/docs/Rakefile:20:in `block in <top (required)>'
Tasks: TOP => gen => init
(See full trace by running task with --trace)
Makefile:79: recipe for target 'gen_docs' failed
make: *** [gen_docs] Error 1
```

Some permission error. Again on creating build directory insid docs, I am again getting permission error for other content as well
```
Step 15/15 : CMD -T
 ---> Using cache
 ---> eae1e8d53293
Successfully built eae1e8d53293
cd docs && docker run -e MINISHIFT_VERSION=1.0.0-beta.5 -tiv /home/budhram/gowork/src/github.com/minishift/minishift/docs:/home/docs:Z minishift/docs gen
mkdir -p build
cp source/_topic_map.yml build/_topic_map.yml
rake aborted!
Errno::EACCES: Permission denied @ rb_sysopen - build/_topic_map.yml
/home/docs/Rakefile:30:in `block in <top (required)>'
Tasks: TOP => gen => markdown_to_asciidoc => build/_topic_map.yml
(See full trace by running task with --trace)
Makefile:79: recipe for target 'gen_docs' failed
make: *** [gen_docs] Error 1
```

File: Makefile [hunk 1]
```diff
@@ -73,15 +72,19 @@ build_docs_container:
 
 .PHONY: gen_adoc_tar
 gen_adoc_tar: synopsis_docs build_docs_container
-	cd docs && docker run -e MINISHIFT_VERSION=$(MINISHIFT_VERSION) -tiv $(shell pwd)/docs:/docs:Z minishift/docs clean adoc_tar
+	cd docs && docker run -e MINISHIFT_VERSION=$(MINISHIFT_VERSION) -tiv $(shell pwd)/docs:/home/docs:Z minishift/docs clean adoc_tar
 
 .PHONY: gen_docs
 gen_docs: synopsis_docs build_docs_container
-	cd docs && docker run -e MINISHIFT_VERSION=$(MINISHIFT_VERSION) -tiv $(shell pwd)/docs:/docs:Z minishift/docs gen
+	cd docs && docker run -e MINISHIFT_VERSION=$(MINISHIFT_VERSION) -tiv $(shell pwd)/docs:/home/docs:Z minishift/docs gen
```
Comment (budhrg) [inline Makefile]: @hferentschik , what this part of docker run does? 
```
v $(shell pwd)/docs:/home/docs:Z minishift/docs gen
```
It definitely add volume and ?

**Comment by hferentschik** [comment]
@budhrg have you manually deleted the _build_ directory using sudo once? If the build directory is still owned by root, a cleanup won't be possible.

File: Makefile (hunk 1 above)
Comment (hferentschik) [inline Makefile]: that's a volume mount. It mounts your local checkout into the container in order to build the docs in there. And yes, _gen_ is the Rakefile target.

**Comment by hferentschik** [comment]
@budhrg can you try without the :Z option in the mount command. I don't think this is really required and might cause some problems as well.

File: docs/Dockerfile [hunk 2]
```diff
@@ -16,12 +16,20 @@ FROM ruby:2.4
 
 RUN apt-get update && apt-get install -y pandoc
 
+ENV UID 1000
```
Comment (budhrg) [inline docs/Dockerfile]: @hferentschik   Why not $USER ? Following changes seems to be working smoothly
```
RUN apt-get update && apt-get install -y pandoc

RUN mkdir -p /home/docs
RUN chown $USER:$USER /home/docs

# From here we run everything as dev user
USER $USER
...
```

Comment (hferentschik) [inline docs/Dockerfile]: And where would $USER come from. It is not defined. 

So you are saying even after once cleaning the _build_ directory, it still does not work?

**Comment by LalatenduMohanty** [comment]
Tested on F25 with native docker runnning. `make gen_docs; make clean_docs` working for me. 
```
$ make clean

$ make
$ make gen_docs 
cd docs && docker build -t minishift/docs .
Sending build context to Docker daemon   129 kB
Step 1 : FROM ruby:2.4
Trying to pull repository docker.io/library/ruby ... 
sha256:52bef497b06d868e77867c95672a6905e95ef3a3ac825c9087224289f852520d: Pulling from docker.io/library/ruby
6d827a3ef358: Pull complete 
2726297beaf1: Pull complete 
7d27bd3d7fec: Pull complete 
44ae682c18a3: Pull complete 
59c43fd2bed9: Pull complete 
64641d23d8a9: Pull complete 
b595e16a8009: Pull complete 
5a4684e04295: Pull complete 
Digest: sha256:52bef497b06d868e77867c95672a6905e95ef3a3ac825c9087224289f852520d
Status: Downloaded newer image for docker.io/ruby:2.4
 ---> d2cee8adb148
Step 2 : RUN apt-get update && apt-get install -y pandoc
 ---> Running in 7925f75b38fe
Get:1 http://security.debian.org jessie/updates InRelease [63.1 kB]
Ign http://deb.debian.org jessie InRelease
Get:2 http://deb.debian.org jessie-updates InRelease [145 kB]
Get:3 http://security.debian.org jessie/updates/main amd64 Packages [457 kB]
Get:4 http://deb.debian.org jessie Release.gpg [2373 B]
Get:5 http://deb.debian.org jessie Release [148 kB]
Get:6 http://deb.debian.org jessie-updates/main amd64 Packages [17.6 kB]
Get:7 http://deb.debian.org jessie/main amd64 Packages [9049 kB]
Fetched 9883 kB in 33min 54s (4856 B/s)
Reading package lists...
Reading package lists...
Building dependency tree...
Reading state information...
The following extra packages will be installed:
  liblua5.1-0 pandoc-data
Suggested packages:
  texlive-latex-recommended texlive-xetex texlive-luatex pandoc-citeproc
  etoolbox
The following NEW packages will be installed:
  liblua5.1-0 pandoc pandoc-data
0 upgraded, 3 newly installed, 0 to remove and 0 not upgraded.
Need to get 4764 kB of archives.
After this operation, 38.9 MB of additional disk space will be used.
Get:1 http://deb.debian.org/debian/ jessie/main liblua5.1-0 amd64 5.1.5-7.1 [108 kB]
Get:2 http://deb.debian.org/debian/ jessie/main pandoc-data all 1.12.4.2~dfsg-1 [202 kB]
Get:3 http://deb.debian.org/debian/ jessie/main pandoc amd64 1.12.4.2~dfsg-1+b14 [4453 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 4764 kB in 4min 53s (16.3 kB/s)
Selecting previously unselected package liblua5.1-0:amd64.
(Reading database ... 21168 files and directories currently installed.)
Preparing to unpack .../liblua5.1-0_5.1.5-7.1_amd64.deb ...
Unpacking liblua5.1-0:amd64 (5.1.5-7.1) ...
Selecting previously unselected package pandoc-data.
Preparing to unpack .../pandoc-data_1.12.4.2~dfsg-1_all.deb ...
Unpacking pandoc-data (1.12.4.2~dfsg-1) ...
Selecting previously unselected package pandoc.
Preparing to unpack .../pandoc_1.12.4.2~dfsg-1+b14_amd64.deb ...
Unpacking pandoc (1.12.4.2~dfsg-1+b14) ...
Setting up liblua5.1-0:amd64 (5.1.5-7.1) ...
Setting up pandoc-data (1.12.4.2~dfsg-1) ...
Setting up pandoc (1.12.4.2~dfsg-1+b14) ...
Processing triggers for libc-bin (2.19-18+deb8u7) ...
 ---> e37853c150c4
Removing intermediate container 7925f75b38fe
Step 3 : ENV UID 1000
 ---> Running in 0ef1515d2683
 ---> ffa45c8f6d79
Removing intermediate container 0ef1515d2683
Step 4 : RUN groupadd -r docs && useradd -g docs -u $UID docs
 ---> Running in 5e9343f51617
 ---> 152c276df409
Removing intermediate container 5e9343f51617
Step 5 : RUN mkdir -p /home/docs
 ---> Running in 5a7ae9a634e3
 ---> 063effa5e334
Removing intermediate container 5a7ae9a634e3
Step 6 : RUN chown docs:docs /home/docs
 ---> Running in 177ba0ccee42
 ---> bc578180f049
Removing intermediate container 177ba0ccee42
Step 7 : USER docs
 ---> Running in 2b90e0a4ac1d
 ---> 6f28344dc19c
Removing intermediate container 2b90e0a4ac1d
Step 8 : ADD Gemfile .
 ---> 837e8c11c442
Removing intermediate container 0ff73316f49c
Step 9 : ADD Gemfile.lock .
 ---> 18cd0e718214
Removing intermediate container bbe4886a5abe
Step 10 : RUN /bin/bash -l -c "bundle install;"
 ---> Running in 3eb18d613b63
Fetching gem metadata from https://rubygems.org/...............
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Installing concurrent-ruby 1.0.5
Installing i18n 0.7.0
Installing minitest 5.10.1
Installing thread_safe 0.3.6
Installing public_suffix 2.0.5
Installing asciidoctor 1.5.5
Installing backports 3.6.8
Installing coffee-script-source 1.12.2
Installing execjs 2.7.0
Installing sass 3.4.23
Installing contracts 0.13.0
Installing dotenv 2.2.0
Installing eventmachine 1.2.3 with native extensions
Installing http_parser.rb 0.6.0 with native extensions
Installing erubis 2.7.0
Installing fast_blank 1.0.0 with native extensions
Installing fastimage 2.1.0
Installing ffi 1.9.18 with native extensions
Installing tilt 2.0.6
Installing hashie 3.5.5
Installing kramdown 1.13.2
Installing libv8 3.16.14.17 (x86_64-linux)
Installing rb-fsevent 0.9.8
Installing memoist 0.15.0
Installing thor 0.19.4
Using bundler 1.14.6
Installing parallel 1.10.0
Installing rack 2.0.1
Installing servolux 0.12.0
Installing rouge 2.0.7
Installing redcarpet 3.4.0 with native extensions
Installing ref 2.0.0
Installing hamster 3.0.0
Installing tzinfo 1.2.2
Installing addressable 2.5.0
Installing coffee-script 2.4.1
Installing uglifier 3.1.5
Installing compass-import-once 1.0.5
Installing em-websocket 0.5.1
Installing rb-inotify 0.9.8
Installing haml 4.0.7
Installing middleman-cli 4.2.1
Installing rack-livereload 0.3.16
Installing therubyracer 0.12.3 with native extensions
Installing activesupport 5.0.2
Installing listen 3.0.8
Installing padrino-support 0.13.3.3
Installing padrino-helpers 0.13.3.3
Installing middleman-core 4.2.1
Installing middleman 4.2.1
Installing middleman-asciidoc 1.0.0.rc.2
Installing middleman-livereload 3.4.6
Installing middleman-syntax 3.0.0
Bundle complete! 8 Gemfile dependencies, 53 gems now installed.
Bundled gems are installed into /usr/local/bundle.
Post-install message from haml:

HEADS UP! Haml 4.0 has many improvements, but also has changes that may break
your application:

* Support for Ruby 1.8.6 dropped
* Support for Rails 2 dropped
* Sass filter now always outputs <style> tags
* Data attributes are now hyphenated, not underscored
* html2haml utility moved to the html2haml gem
* Textile and Maruku filters moved to the haml-contrib gem

For more info see:

http://rubydoc.info/github/haml/haml/file/CHANGELOG.md

 ---> 9833175d2374
Removing intermediate container 3eb18d613b63
Step 11 : WORKDIR /home/docs
 ---> Running in cd013feda7f9
 ---> 75ee8e1962d9
Removing intermediate container cd013feda7f9
Step 12 : EXPOSE 35729:35729
 ---> Running in b7989deed7f8
 ---> efdbd8d01638
Removing intermediate container b7989deed7f8
Step 13 : EXPOSE 4567:4567
 ---> Running in 653f9639f8a8
 ---> 60c07ac90bbe
Removing intermediate container 653f9639f8a8
Step 14 : ENTRYPOINT rake
 ---> Running in ec199ff32ad2
 ---> d9abf3e8f739
Removing intermediate container ec199ff32ad2
Step 15 : CMD -T
 ---> Running in 90e52d9d4ecf
 ---> f4df70bdd52b
Removing intermediate container 90e52d9d4ecf
Successfully built f4df70bdd52b
cd docs && docker run -e MINISHIFT_VERSION=1.0.0-beta.5 -tiv /home/lmohanty/GoCode/src/github.com/minishift/minishift/docs:/home/docs:Z minishift/docs gen
mkdir -p build
cp source/_topic_map.yml build/_topic_map.yml
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift.adoc source/_tmp/minishift.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_config.adoc source/_tmp/minishift_config.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_config_get.adoc source/_tmp/minishift_config_get.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_config_set.adoc source/_tmp/minishift_config_set.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_config_unset.adoc source/_tmp/minishift_config_unset.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_config_view.adoc source/_tmp/minishift_config_view.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_console.adoc source/_tmp/minishift_console.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_delete.adoc source/_tmp/minishift_delete.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_docker-env.adoc source/_tmp/minishift_docker-env.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_ip.adoc source/_tmp/minishift_ip.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_logs.adoc source/_tmp/minishift_logs.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift.adoc source/_tmp/minishift_openshift.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_config.adoc source/_tmp/minishift_openshift_config.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_config_set.adoc source/_tmp/minishift_openshift_config_set.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_config_view.adoc source/_tmp/minishift_openshift_config_view.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_list-versions.adoc source/_tmp/minishift_openshift_list-versions.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_restart.adoc source/_tmp/minishift_openshift_restart.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_service.adoc source/_tmp/minishift_openshift_service.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_openshift_service_list.adoc source/_tmp/minishift_openshift_service_list.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_ssh.adoc source/_tmp/minishift_ssh.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_start.adoc source/_tmp/minishift_start.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_status.adoc source/_tmp/minishift_status.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_stop.adoc source/_tmp/minishift_stop.md
pandoc --atx-headers --to=asciidoc --output=source/command-ref/minishift_version.adoc source/_tmp/minishift_version.md
bundle exec middleman build
Warning: `set :markdown_engine` is deprecated. Use `config[:markdown_engine] =` instead.
Warning: `set :markdown` is deprecated. Use `config[:markdown] =` instead.
asciidoctor: WARNING: index.adoc: line 3: include file not found: /home/docs/source/minishift/variables.adoc
      create  build/stylesheets/site.css
      create  build/images/icons/note.png
      create  build/javascripts/site.js
asciidoctor: WARNING: <stdin>: line 3: include file not found: /home/docs/source/minishift/variables.adoc
asciidoctor: ERROR: <stdin>: line 6: only book doctypes can contain level 0 sections
      create  build/index.html
      create  build/getting-started/index.html
      create  build/images/.keep
      create  build/getting-started/docker-machine-drivers.html
      create  build/getting-started/.keep
      create  build/developing/index.html
      create  build/using/troubleshooting.html
      create  build/developing/developing.html
      create  build/developing/writing-docs.html
      create  build/using/index.html
      create  build/command-ref/minishift_start.html
      create  build/command-ref/minishift_delete.html
      create  build/using/interacting-with-openshift.html
      create  build/command-ref/minishift_console.html
      create  build/command-ref/minishift_ip.html
      create  build/using/managing-minishift.html
      create  build/command-ref/minishift_openshift_service_list.html
      create  build/command-ref/minishift_openshift_config_view.html
      create  build/command-ref/minishift_config.html
      create  build/command-ref/minishift_stop.html
      create  build/command-ref/minishift.html
      create  build/command-ref/minishift_version.html
      create  build/command-ref/.keep
      create  build/command-ref/minishift_openshift.html
      create  build/command-ref/minishift_ssh.html
      create  build/command-ref/minishift_docker-env.html
      create  build/command-ref/minishift_status.html
      create  build/command-ref/minishift_config_set.html
      create  build/command-ref/minishift_logs.html
      create  build/command-ref/minishift_openshift_list-versions.html
      create  build/command-ref/minishift_config_view.html
      create  build/command-ref/minishift_openshift_config_set.html
      create  build/command-ref/minishift_config_get.html
      create  build/command-ref/minishift_openshift_restart.html
      create  build/command-ref/minishift_openshift_config.html
      create  build/command-ref/minishift_config_unset.html
      create  build/command-ref/minishift_openshift_service.html
      remove  build/_topic_map.yml
      remove  build/variables.adoc
Project built successfully.
[lmohanty@LalatenduM-laptop minishift]$ make clean_docs
cd docs && docker build -t minishift/docs .
Sending build context to Docker daemon   364 kB
Step 1 : FROM ruby:2.4
 ---> d2cee8adb148
Step 2 : RUN apt-get update && apt-get install -y pandoc
 ---> Using cache
 ---> e37853c150c4
Step 3 : ENV UID 1000
 ---> Using cache
 ---> ffa45c8f6d79
Step 4 : RUN groupadd -r docs && useradd -g docs -u $UID docs
 ---> Using cache
 ---> 152c276df409
Step 5 : RUN mkdir -p /home/docs
 ---> Using cache
 ---> 063effa5e334
Step 6 : RUN chown docs:docs /home/docs
 ---> Using cache
 ---> bc578180f049
Step 7 : USER docs
 ---> Using cache
 ---> 6f28344dc19c
Step 8 : ADD Gemfile .
 ---> Using cache
 ---> 837e8c11c442
Step 9 : ADD Gemfile.lock .
 ---> Using cache
 ---> 18cd0e718214
Step 10 : RUN /bin/bash -l -c "bundle install;"
 ---> Using cache
 ---> 9833175d2374
Step 11 : WORKDIR /home/docs
 ---> Using cache
 ---> 75ee8e1962d9
Step 12 : EXPOSE 35729:35729
 ---> Using cache
 ---> efdbd8d01638
Step 13 : EXPOSE 4567:4567
 ---> Using cache
 ---> 60c07ac90bbe
Step 14 : ENTRYPOINT rake
 ---> Using cache
 ---> d9abf3e8f739
Step 15 : CMD -T
 ---> Using cache
 ---> f4df70bdd52b
Successfully built f4df70bdd52b
cd docs && docker run -e MINISHIFT_VERSION=1.0.0-beta.5 -tiv /home/lmohanty/GoCode/src/github.com/minishift/minishift/docs:/home/docs:Z minishift/docs clean
rm -f source/command-ref/minishift.adoc source/command-ref/minishift_config.adoc source/command-ref/minishift_config_get.adoc source/command-ref/minishift_config_set.adoc source/command-ref/minishift_config_unset.adoc source/command-ref/minishift_config_view.adoc source/command-ref/minishift_console.adoc source/command-ref/minishift_delete.adoc source/command-ref/minishift_docker-env.adoc source/command-ref/minishift_ip.adoc source/command-ref/minishift_logs.adoc source/command-ref/minishift_openshift.adoc source/command-ref/minishift_openshift_config.adoc source/command-ref/minishift_openshift_config_set.adoc source/command-ref/minishift_openshift_config_view.adoc source/command-ref/minishift_openshift_list-versions.adoc source/command-ref/minishift_openshift_restart.adoc source/command-ref/minishift_openshift_service.adoc source/command-ref/minishift_openshift_service_list.adoc source/command-ref/minishift_ssh.adoc source/command-ref/minishift_start.adoc source/command-ref/minishift_status.adoc source/command-ref/minishift_stop.adoc source/command-ref/minishift_version.adoc
rm -rf build
```

**Comment by budhrg** [comment]
Merged and Thanks!

**Pull request closed without merging**